pkg/api/dirs.go

```diff
@@ -0,0 +1,149 @@
+	toEncrypt bool
+	reader    io.ReadCloser
+}
+
+// dirUploadHandler uploads a directory supplied as a tar in an HTTP request
+func (s *server) dirUploadHandler(w http.ResponseWriter, r *http.Request) {
+	dirInfo, err := getDirHTTPInfo(r)
+	if err != nil {
+		s.Logger.Errorf("dir upload, get dir info: %v", err)
+		jsonhttp.BadRequest(w, "could not extract dir info from request")
+		return
+	}
+
+	reference, err := storeDir(r.Context(), dirInfo, s.Storer, s.Logger)
+	if err != nil {
+		s.Logger.Errorf("dir upload, store dir: %v", err)
+		jsonhttp.InternalServerError(w, "could not store dir")
+		return
+	}
+
+	jsonhttp.OK(w, fileUploadResponse{
+		Reference: reference,
+	})
+}
+
+// getDirHTTPInfo extracts data for a directory to be uploaded from an HTTP request
+func getDirHTTPInfo(r *http.Request) (*dirUploadInfo, error) {
+	if r.Body == http.NoBody {
+		return &dirUploadInfo{}, errors.New("request has no body")
+	}
+	toEncrypt := strings.ToLower(r.Header.Get(EncryptHeader)) == "true"
+	return &dirUploadInfo{
+		reader:    r.Body,
+		toEncrypt: toEncrypt,
+	}, nil
+}
+
+// storeDir stores all files recursively contained in the directory given as a tar
+// it returns the hash for the uploaded manifest corresponding to the uploaded dir
+func storeDir(ctx context.Context, dirInfo *dirUploadInfo, s storage.Storer, logger logging.Logger) (swarm.Address, error) {
+	dirManifest := jsonmanifest.NewManifest()
+
+	// set up HTTP body reader
+	tarReader := tar.NewReader(dirInfo.reader)
+	defer dirInfo.reader.Close()
+
+	// iterate through the files in the supplied tar
+	for {
+		fileHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("read tar stream error: %v", err)
+		}
+
+		filePath := fileHeader.Name
+
+		// only store regular files
+		if !fileHeader.FileInfo().Mode().IsRegular() {
+			logger.Warningf("skipping file upload for %s as it is not a regular file", filePath)
+			continue
+		}
+
+		fileName := fileHeader.FileInfo().Name()
+		contentType := mime.TypeByExtension(filepath.Ext(fileHeader.Name))
+
+		// upload file
+		fileInfo := &fileUploadInfo{
+			name:        fileName,
+			size:        fileHeader.FileInfo().Size(),
+			contentType: contentType,
+			toEncrypt:   dirInfo.toEncrypt,
+			reader:      tarReader,
+		}
+		fileReference, err := storeFile(ctx, fileInfo, s)
+		if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("store dir file error: %v", err)
+		}
+		logger.Tracef("uploaded dir file %v with reference %v", filePath, fileReference)
+
+		// create manifest entry for uploaded file
+		headers := http.Header{}
+		headers.Set("Content-Type", contentType)
+		fileEntry := &jsonmanifest.JSONEntry{
+			Reference: fileReference,
+			Name:      fileName,
+			Headers:   headers,
+		}
+
+		// add entry to dir manifest
+		dirManifest.Add(filePath, fileEntry)
+	}
+
+	// check if files were uploaded through the manifest
+	if len(dirManifest.Entries) == 0 {
```

Comment:
oops. you're accessing an exposed map field which is not protected. so this implementation is not thread safe. would be better if jsonmanifest does not expose this field and has designated getters with mutexes protecting field access. obviously this is not an issue right now since the implementation is single threaded. still better to have this as thread safe from the get-go

Reply:
nice catch! added this to #480

Reply:
As it was mentioned, this `jsonmanifest` implementation was initially created for tests around this feature, and after some iteration decided to be left as first version. This was all while trying to be in accordance with [feature notes](https://hackmd.io/YklWpLHPRDGWipoZavw50Q).
If we are to have full-blown JSON implementation for this, there are multiple things that need to be changed. Not just for manifest structure, but around entry one also.

Regarding protecting map access with mutex, that is easy, however not sure if there are some cases where we would have the case where this is needed (at least for now). If the policy is to always have a mutex when there is map in external struct, that is okay, and I will follow this going forward.